### PR TITLE
Closes #514

### LINKED ISSUE #514: Cleanup: `maxDose` method for `IncrementsRelativeParts` signature.
Please edit appropriately below if needed.

To do:

- [x] Revise documentation
  - [x] RStudio outline section dividers (use `# function or class name ----`)
  - [x] Add lifecycle badge (as a default take "stable" for existing functions or classes)
  - [x] Replace `##'` by `#'`
  - [x] Add to `pkgdown.yml`
  - [x] Remove file history and the timestamp in comments on top of file
- [x] Add unit tests
  - [x] For classes:
    - [x] Test default constructor
    - [x] For reference classes: test methods
  - [x] For functions: as usual
- [x] Refactor as needed
  - [x] Code styling (indentation etc.) - install package `styler` and use corresponding add-in
  - [x] Keep externally used names so that we don't break downstream code
  - [x] Remove `keywords` tags
  - [x] For functions or methods:
    - [x] Add assertions for arguments using `checkmate` functions
    - [x] If it is too long, break up into pieces (rule of thumb: maximum 50 lines of code)
    - [x] Consider renaming internal variables etc. if needed
    - [x] Avoid explicit use of `return`
  - [x] For classes:
    - [x] move function definitions outside if possible (e.g. move validation function as `validity_subjects()` to `Data-validity.R` and have corresponding test file `test-Data-validity.R`)
    - [x] In `setClass` function, just use the `slots` instead of outdated `representation`
    - [x] Remove things like `validObject(.Data())`

### PR DESCRIPTION
**Pull request by wwojciech**
Closes #514 - draft, no unit tests

### CODE REVIEW

**Comment by wwojciech** [comment]
Hi @danielinteractive - I am submitting this PR as a draft (i.e. all ready except unit tests) since the logic of `maxDose-IncrementsRelativeParts` function is quite involved, it was not consistent with the help page for `IncrementsRelativeParts` class. So, cleaned it (also the help fro `IncrementsRelativeParts`) according to my best intuition and knowledge. Before I proceed with the unit tests, I wanted to ask you to quickly have a look at this and confirm if this is ok and I can proceed to the unit tests. Tx!

File: R/Rules-methods.R [hunk 1]
```diff
@@ -1138,71 +1138,55 @@ setMethod(
...
-      increments = "IncrementsRelativeParts",
-      data = "DataParts"
-    ),
-  def =
-    function(increments, data, ...) {
-      ## determine if there are already cohorts
-      ## belonging to part 2:
-      alreadyInPart2 <- any(data@part == 2L)
+    # If there are already cohorts belonging to part 2.
+    if (any(data@part == 2L)) {
```
Comment (danielinteractive) [inline R/Rules-methods.R:1160]: I would keep the variable assignment here, since that explains what it is, even without the comment

**Review by danielinteractive — COMMENTED** [review]
looks good, thanks Wojciech

**Comment by wwojciech** [comment]
Hi @danielinteractive - I think that here:
https://github.com/Roche/crmPack/blob/f041ae22d3991892257eaea30840ddbc804c73ac/R/Rules-methods.R#L1187
we need to extract the case when:
```
lastDoseLevel + increments@dlt_start >= length(data@part1Ladder)
```
I think in such a case we should call `callNextMethod(increments, data, ...)`, i.e. a usual relative increment rules apply.

What do you think?

**Comment by danielinteractive** [comment]
> Hi @danielinteractive - I think that here:
> 
> https://github.com/Roche/crmPack/blob/f041ae22d3991892257eaea30840ddbc804c73ac/R/Rules-methods.R#L1187
> 
> we need to extract the case when:
> 
> ```
> 
> (lastDoseLevel + increments@dlt_start) >= length(data@part1Ladder)
> 
> ```
> 
> I think in such a case we should call `callNextMethod(increments, data, ...)`, i.e. a usual relative increment rules apply.
> 
> 
> 
> What do you think?

Thanks Wojciech, I think we cannot refer to the parent method because we still need to use the ladder slot to select the dose?

**Comment by wwojciech** [comment]
Hi @danielinteractive, Could you pl. accept this PR if all looks ok? Tx!

**Comment by wwojciech** [comment]
> > > Hi @danielinteractive - I think that here:
> > > https://github.com/Roche/crmPack/blob/f041ae22d3991892257eaea30840ddbc804c73ac/R/Rules-methods.R#L1187
> > > 
> > > we need to extract the case when:
> > > ```
> > > 
> > > (lastDoseLevel + increments@dlt_start) >= length(data@part1Ladder)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I think in such a case we should call `callNextMethod(increments, data, ...)`, i.e. a usual relative increment rules apply.
> > > What do you think?
> > 
> > 
> > Thanks Wojciech, I think we cannot refer to the parent method because we still need to use the ladder slot to select the dose?
> 
> yap, but how can we use it since we are already exceeding `part1Ladder` "scale", and also we are already in part2....

I created a separate issue ( #517 ) to clarify this, and will close this PR now. Thanks!

**Comment by pengguanya** [comment]
@danielinteractive I just found this probably has been implemented already:
https://github.com/Roche/crmPack/blob/8a7cffa2a0263d008f321311bc615d4d64b28960/R/Rules-methods.R#L1567

Please confirm if we should close this issue.

**Pull request merged**